File: problems/1143/1143.longest-common-subsequence.py
#
# @lc app=leetcode id=1143 lang=python3
#
# [1143] Longest Common Subsequence
#

# @lc code=start
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n, m = len(text1), len(text2)
        dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]
        
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if text1[i - 1] != text2[j - 1]:
                    dp[j][i] = max(dp[j - 1][i], dp[j][i - 1])
                else:
                    dp[j][i] = dp[j - 1][i - 1] + 1
        
        return dp[-1][-1]
        
# @lc code=end

File: problems/1143/1143.longest-common-subsequence.java
/*
 * @lc app=leetcode id=1143 lang=java
 *
 * [1143] Longest Common Subsequence
 */

// @lc code=start
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        /*
         * return length of text1 and text2 longest common subsequence.
         * if no, return 0.
         *
         * intuition:
         * - bottom-up dp
         * - if text1[i] = text2[j] -> dp(i,j) = 1 + dp(i+1,j+1)
         * - else -> dp(i,j) = max(dp(i+1,j), dp(i,j+1)) skip i or skip j
         *
         * O(N * M) time, O(N * M) space
         */

        int n = text1.length();
        int m = text2.length();
        int[][] dp = new int[n + 1][m + 1];

        for (int i = n - 1; i >= 0; i--) {
            for (int j = m - 1; j >= 0; j--) {
                if (text1.charAt(i) == text2.charAt(j)) {
                    dp[i][j] = 1 + dp[i + 1][j + 1];
                } else {
                    dp[i][j] = Math.max(dp[i + 1][j], dp[i][j + 1]);
                }
            }
        }

        return dp[0][0];
    }
}
// @lc code=end
